Documentation/teaching/lectures/smp.rst

```diff
@@ -0,0 +1,1156 @@
+architecture independent APIs to disable and enable interrupts:
+
+.. slide:: Synchronization with interrupts (x86)
+   :inline-contents: True
+   :level: 2
+
+   .. code-block:: c
+
+       #define local_irq_disable() \
+           asm volatile („cli” : : : „memory”)
+
+      #define local_irq_enable() \
+          asm volatile („sti” : : : „memory”)
+
+      #define local_irq_save(flags) \
+          asm volatile ("pushf ; pop %0" :"=g" (flags)
+                        : /* no input */: "memory")
+
+      #define local_irq_restore(flags) \
+          asm volatile ("push %0 ; popf"
+                        : /* no output */
+                        : "g" (flags) :"memory", "cc");
+
+
+Although the interrupts can be explicitly disabled and enable with
+:c:func:`local_irq_disable` and :c:func:`local_irq_enable` these APIs
+should only be used when the current state and interrupts is
+known. They are usually used in core kernel code (like interrupt
+handling).
+
+For typical cases where we want to avoid interrupts due to concurrency
+issues it is recommended to use the :c:func:`local_irq_save` and
+:c:func:`local_irq_restore` variants. They take care of saving and
+restoring the interrupts states so they can be freely called from
+overlapping critical sections without the risk of accidentally
+enabling interrupts while still in a critical section, as long as the
+calls are balanced.
+
+Spin Locks
+==========
+
+Spin locks are used to serialize access to a critical section. They
+are necessary on multi-core systems where we can have true execution
+parallelism. This is a typical spin lock implementation:
+
+
+.. slide:: Spin Lock Implementation Example (x86)
+   :inline-contents: True
+   :level: 2
+
+   .. code-block:: asm
+
+      spin_lock:
+          lock bts [my_lock], 0
+	  jc spin_lock
+
+      /* critical section */
+
+      spin_unlock:
+          mov [my_lock], 0
+
+   **bts** - bit test and set; it copies the src bit from the dts
+   memory address to the carry flag and then sets it
+
+
+As it can be seen, the spin lock uses an atomic instruction to make
+sure that only one core can enter the critical section. If there are
+multiple cores trying to enter they will continuously "spin" until the
+lock is released.
+
+While the spin lock avoids race conditions, it can have a significant
+impact on the system's performance due to "lock contention":
+
+
+.. slide:: Lock Contention
+   :inline-contents: True
+   :level: 2
+
+   * There is lock contention when at least one core spins trying to
+     enter the critical section lock
+
+   * Lock contention grows with the critical section size, time spent
+     in the critical section and the number of cores in the system
+
+
+Another negative side effect of spin locks is cache thrashing.
+
+.. slide:: Cache Thrashing
+   :inline-contents: True
+   :level: 2
+
+   Cache thrashing occurs when multiple cores are trying to read and
+   write to the same memory resulting in excessive cache misses.
```

Comment:
Add "When one core writes, the others have to invalidate its cache lines and then populate them (after a cache miss) with the updated values from the main memory, incurring overhead."